Make lazy tokens lazier

SquidDev opened the pull request:
- Defer reading the user-supplied classpath until needed.
 - Pass around `FileCollection`s instead of `Set<File>`s, so we don't need to resolve the contents until actually needed.

SquidDev committed:
Make lazy tokens lazier

 - Defer reading the user-supplied classpath until needed.
 - Pass around FileCollections instead of Set<File>s, so we don't need
   to resolve the contents until actually needed.

marchermans commented:
I support this fully. But for documentation reasons: why is this change needed?; Other then to comply with gradles examples that is....

SquidDev commented:
I've been fiddling with a couple of alternative approaches to #924, as from conversations on Discord, people weren't very happy with that PR.

I ended up having a few issues where we end up trying to evaluate tokens when configuring a task, which causes deobf dependencies to be fetched, which in turn causes all sorts of other problems, and so it felt worth fixing that first before going any further.

XFactHD commented on src/common/java/net/minecraftforge/gradle/common/util/runs/GenIDERunsTask.java:
```diff
@@ -41,11 +40,9 @@ public void run() {
 
         runConfigGenerator.createRunConfiguration(minecraft, runConfigurationsDir, project,
-                additionalClientArgs, minecraftArtifacts, runtimeClasspathArtifacts);
+                additionalClientArgs, getMinecraftArtifacts(), getMinecraftArtifacts());
```
This doesn't look right. Shouldn't it be `this.getMinecraftArtifacts(), this.getRuntimeClasspathArtifacts()` instead of `this.getMinecraftArtifacts(), this.getMinecraftArtifacts()`?

SquidDev replied:
Good spot, thanks!

SquidDev committed:
Fix gen runs using the wrong classpath

SizableShrimp commented on src/common/java/net/minecraftforge/gradle/common/util/runs/RunConfigGenerator.java:
```diff
@@ -181,6 +174,15 @@ protected static Map<String, Supplier<String>> configureTokensLazy(final Project
+            String resolvedClasspath = getResolvedClasspath(classpath.getFiles());
+            if (supplier == null) return resolvedClasspath;
+            String oldCp = supplier.get();
+            return oldCp == null || oldCp.isEmpty() ? resolvedClasspath : String.join(File.pathSeparator, oldCp, resolvedClasspath);
```
This should probably be improved to be an optional merge if either classpath string is empty or null to return the one that is non-null. I don't think our file collection will ever be empty, but it is best to safeguard anyways. And if both are null or empty perhaps return an empty string or null string depending on compute()'s contract?

SquidDev replied:
Yep, that's a good point. I think we want to avoid returning `null` here, as `classpathFileWriter` assumes the returned string is non-null.

SizableShrimp commented on src/common/java/net/minecraftforge/gradle/common/util/runs/GenIDERunsTask.java:
```diff
@@ -41,11 +40,9 @@ public void run() {
 
         runConfigGenerator.createRunConfiguration(minecraft, runConfigurationsDir, project,
-                additionalClientArgs, minecraftArtifacts, runtimeClasspathArtifacts);
+                additionalClientArgs, getMinecraftArtifacts(), getRuntimeClasspathArtifacts());
```
Consistency: The other class has a `this.` prefix you kept, while you removed it here. Probably add it.

SizableShrimp reviewed (changes requested):
This looks fine. My comments are minor. I will likely test it in a few days when I am able to.

SquidDev committed:
Review cleanup

 - Add back this-s for consistency
 - Make classpath merging in makeClasspathToken a little more symmetric.

SizableShrimp commented on src/common/java/net/minecraftforge/gradle/common/util/runs/RunConfigGenerator.java:
```diff
@@ -179,7 +179,9 @@ private static BiFunction<String, Supplier<String>, Supplier<String>> makeClassp
-            return oldCp == null || oldCp.isEmpty() ? resolvedClasspath : String.join(File.pathSeparator, oldCp, resolvedClasspath);
+            if (Strings.isNullOrEmpty(oldCp)) return resolvedClasspath;
+            if (Strings.isNullOrEmpty(resolvedClasspath)) return oldCp;
+            return String.join(File.pathSeparator, oldCp, resolvedClasspath);
```
How does this guarantee that an empty string is returned if both are null-or-empty? Doesn't look like it does.

SquidDev replied:
```java
// resolvedClasspath != null by definition of getResolvedClasspath.
if (Strings.isNullOrEmpty(oldCp)) return resolvedClasspath;
// At this point, oldCp != null, so this returns a non-null value.
if (Strings.isNullOrEmpty(resolvedClasspath)) return oldCp;
// Finally, oldCp != null && resolvedClasspath != null
return String.join(File.pathSeparator, oldCp, resolvedClasspath);
```

I can add a comment to clarify if that would be helpful?

SizableShrimp replied:
Ah, that makes sense. Perhaps just add a `@NotNull` annotation to getResolvedClasspath, that should be good enough for the future.

SquidDev replied:
:+1: I've gone for `@Nonnull` as the rest of FG seems to use javax annotations right now, but can change if preferred.

SquidDev committed:
Annotate getResolvedClasspath as non-null

SizableShrimp reviewed (approved):
If no one else on the team gets to this, I will probably test it in a few days when I'm at a PC. Cc @marchermans if he wants to test it

SquidDev commented:
Thanks!
